### PR TITLE
create model based on pre-trained VGG16 (WIP)

### PR DESCRIPTION
**Pull request by Temebe**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
 venv/
-Data/
\ No newline at end of file
+Data/
+*.env
+settings.json
```
Comment (Szuumii) [inline .gitignore:4]: Why .json file ?! O.o

File: models.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        super().__init__()
+        self.dataset_csv_path = dataset_csv_path
+        self.dataset_dir_path = dataset_dir_path
+        self.batch_size = batch_size
+
+        self.net = models.vgg16(pretrained=True)
+        last_in_features = self.net.classifier[6].in_features
+        self.net.classifier[6] = nn.Linear(last_in_features, 19)
+
+        self.loss = nn.NLLLoss()
```
Comment (Szuumii) [inline models.py:19]: We should go for cross-entropy loss: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

File: models.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    def forward(self, x):
+        return self.net(x)
+
+    def training_step(self, batch, _):
+        x, _ = batch
```
Comment (Szuumii) [inline models.py:25]: We like to read prod_year instead of _

**Review by Szuumii — CHANGES_REQUESTED** [review]

File: models.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+class DVMModel(pl.LightningModule):
+    def __init__(self, dataset_csv_path, dataset_dir_path, batch_size=32):
+        super().__init__()
+        self.dataset_csv_path = dataset_csv_path
+        self.dataset_dir_path = dataset_dir_path
+        self.batch_size = batch_size
+
+        self.net = models.vgg16(pretrained=True)
+        last_in_features = self.net.classifier[6].in_features
+        self.net.classifier[6] = nn.Linear(last_in_features, 19)
```
[the remark this answers is not in the document]
Comment (Temebe) [reply models.py:17]: That is actually ` Prod_year_max - prod_year_min` :) Just hidden behind magic number. I will clarify this in code.

[the remark this answers is not in the document]
Comment (Szuumii) [reply models.py:17]: Magic numbers are never a good idea :D We should also consider number of classes that are in the main data folder, since they might not be included in this subset of confirmed

File: models.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
+from torchvision import models, transforms
+from torch.utils.data import DataLoader
+import pytorch_lightning as pl
+import torch
+from dataset.CarFronts import FrontDataset
+from torch import nn
+
+class DVMModel(pl.LightningModule):
+    def __init__(self, dataset_csv_path, dataset_dir_path, batch_size=32):
```
[the remark this answers is not in the document]
Comment (Temebe) [reply models.py]: I've added small_train flag to model constructor. It picks randomly 5% of original dataset. Let me know if that is sufficient or if we want to go with specifically prepared sets.

File: models.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+class DVMModel(pl.LightningModule):
+    def __init__(self, dataset_csv_path, dataset_dir_path, batch_size=32, learning_rate=0.0002, input_size=(300, 300), small_train=False):
+        super().__init__()
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.small_train = small_train
+        self.input_size = input_size
+
+        self.dataset = FrontDataset(dataset_csv_path, dataset_dir_path)
+        PROD_YEAR_RANGE = self.dataset.maximum_prod_year() - self.dataset.minimum_prod_year() + 1
```
Comment (Temebe) [inline models.py:17]: This right here is something I am not really comfortable with :)
I get minimum and maximum years of production from csv files, which are processed by FrontDataset class. However, it is needed when we create network, as we are changing outputs of last layer based on these years. I moved creating dataset object in `__init__` function from `prepare_data`, but I am not entirely sure if it is ok.

File: models.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    def training_step(self, batch, _):
+        x, prod_year = batch
+        predictions = self.forward(x)
+        loss = self.loss(predictions, (prod_year - self.dataset.minimum_prod_year()))
+        self.log("train_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
+        return loss
+    
+    def validation_step(self, batch, batch_idx):
+        return self.training_step(batch, batch_idx)
```
Comment (Temebe) [inline models.py:45]: Do we want to create separate folders for validation? Or will we just take some subset of original dataset (randomly or simply by taking last few percentages)?

File: dataset/CarFronts.py [hunk 8]
```diff
@@ -42,6 +41,15 @@ def load_image(self, path):
...
         print(f"Dataset contains {len(self.data)} cars")
 
+    def minimum_prod_year(self):
+        return np.amin(self.data, 0)[1]
+
+    def maximum_prod_year(self):
+        return np.amax(self.data, 0)[1]
+
+    def shrink_randomly(self, size):
+        self.data = self.data[np.random.choice(self.data.shape[0], size, replace=False), :]
```
Comment (Temebe) [inline dataset/CarFronts.py:51]: I came up with this as a way to pick subset of data with flag small_train. Let me know if you have some comments, if it is not readable or if I should simply get rid of this.

File: models.py (hunk 7 above)
Comment (Szuumii) [reply models.py:45]: We could split original csv into test, data, and validation up front. In this case data preparation could be a little more transparent. I'll look for an example and give you feedback later on ;)

**Review by Szuumii — CHANGES_REQUESTED** [review]

File: models.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self, dataset_csv_path, dataset_dir_path, batch_size=32, learning_rate=0.0002, input_size=(300, 300), small_train=False):
+        super().__init__()
+        self.dataset_csv_path = dataset_csv_path
+        self.dataset_dir_path = dataset_dir_path
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.small_train = small_train
+        self.input_size = input_size
+
+        PROD_YEAR_RANGE = FrontDataset.prod_years_range(csv_path)
```
Comment (Temebe) [inline models.py:18]: I came up with static method in FrontDataset, which calculates this range. I was hesitant to putting logic of operating on this csv file in Model class. I am not sure however if we want to keep it as static method. Let me know :)

Comment (Szuumii) [reply models.py:18]: This solution looks great! 👍  However, I still we should ask about it during the consultation

File: models.py [hunk 10]
```diff
@@ -27,18 +28,35 @@ def forward(self, x):
         return self.net(x)
 
     def prepare_data(self):
+        SMALL_DATA_PERCENTAGE = 0.01
```
Comment (Szuumii) [inline models.py:31]: This is correct, however, I think that instead of such a huge chunk of code that creates those sets out random, I'd just give paths to preprocessed CSV files that are split out random. I'll correct it

**Review by Szuumii — APPROVED** [review]

File: model/models.py [hunk 11]
```diff
@@ -28,43 +27,44 @@ def forward(self, x):
...
         return loss
     
     def validation_step(self, batch, batch_idx):
-        return self.training_step(batch, batch_idx)
-
-    def configure_optimizers(self):
-        return torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+        x, prod_year = batch
+        predictions = self.forward(x)
+        loss = self.loss(predictions, (prod_year - self.train_dataset.minimum_prod_year()))
```
Comment (Temebe) [inline model/models.py:66]: Maybe we should set minimum prod year as separate variable in prepare_data method? Using train_dataset in validation_step might be confusing. 

It could be also just predefined like PROD_YEAR_RANGE
